text/0517-io-os-reform.md

```diff
@@ -1370,6 +1370,91 @@ The contents of `std::io::net` submodules `tcp`, `udp`, `ip` and
 the other modules are being moved or removed and are described
 elsewhere.
 
+#### InetAddr
+
+The composition of an `IpAddr` and a port. It has the following interface:
+
+```rust
+impl InetAddr {
+  /// Returns a new InetAddr composed of an unspecified v4 IP and a 0
+  /// port
+  fn any_v4() -> InetAddr;
+
+  /// Returns a new InetAddr composed of an unspecified v6 IP and a 0
+  /// port
+  fn any_v6() -> InetAddr;
+
+  fn ip(&self) -> IpAddr;
+  fn port(&self) -> u16;
+
+  /// Returns true if the IpAddr is unspecified and port == 0
+  fn is_unspecified(&self) -> bool;
+}
+```
+
+#### IpAddr
+
+Represents an IP address. It has the following interface:
+
+```rust
+impl IpAddr {
+  fn new_v4(a: u8, b: u8, c: u8, d: u8) -> IpAddr;
+  fn any_v4() -> IpAddr;
+
+  fn new_v6(a: u16, b: u16, c: u16, d: u16, e: u16, f: u16, g: u16, h: u16) -> IpAddr;
+  fn any_v6() -> IpAddr;
+
+  // The following functions proxy to the versioned IP address value
+  fn is_unspecified(&self) -> bool;
+  fn is_loopback(&self) -> bool;
+  fn is_global(&self) -> bool;
+  fn is_private(&self) -> bool;
+  fn is_multicast(&self) -> bool;
+}
+```
+
+#### Ipv4Addr
+
+Represents a version 4 IP address. It has the following interface:
+
+```rust
+impl Ipv4Addr {
+  fn new(a: u8, b: u8, c: u8, d: u8) -> Ipv4Addr;
+  fn any() -> Ipv4Addr;
+  fn octets(&self) -> [u8; 4];
+  fn is_unspecified(&self) -> bool;
+  fn is_loopback(&self) -> bool;
+  fn is_private(&self) -> bool;
+  fn is_link_local(&self) -> bool;
+  fn is_global(&self) -> bool;
+  fn is_multicast(&self) -> bool;
+  fn to_ipv6_compatible(&self) -> Ipv6Addr;
+  fn to_ipv6_mapped(&self) -> Ipv6Addr;
+}
+```
+
+#### Ipv6Addr
+
+Represents a version 6 IP address. It has the following interface:
+
+```rust
+impl Ipv6Addr {
+  fn new(a: u16, b: u16, c: u16, d: u16, e: u16, f: u16, g: u16, h: u16) -> Ipv6Addr;
```

Comment:
Only that this made it easier for users to switch from the old enum-based `IpAddr::Ipv6Addr`. Otherwise they'd need to do more than a simple addition of `::new`. But that's not a good enough reason, and I think making callers pack the integers into a slice (just add brackets?) is better than making callers who already have a slice break that slice into separate arguments. So I'd change it to `[u16; 8]`, as you said.